### DISCUSSION TITLE
Gunicorn 21.2.0 has been released

### DISCUSSION THREAD

**Discussion by benoitc** [benoitc/gunicorn#3036] [Announcements]
**Gunicorn 21.2.0 has been released.** This version fix the issue introduced in the threaded worker.

Changes:

```
21.2.0 - 2023-07-19
===================
fix thread worker: revert change considering connection as idle .

*** NOTE ***

This is fixing the bad file description error.
```

1. Documentation is available there: https://docs.gunicorn.org/en/stable/news.html
2. Packages: https://pypi.org/project/gunicorn/

<hr /><em>This discussion was created from the release <a href='https://github.com/benoitc/gunicorn/releases/tag/21.2.0'>Gunicorn 21.2.0 has been released</a>.</em>

**Comment by benoitc** [comment]
```
(env311) ➜  examples git:(master) gunicorn -w3 --threads 4 echo:app
[2023-07-19 13:55:59 +0200] [23570] [INFO] Starting gunicorn 21.2.0
[2023-07-19 13:55:59 +0200] [23570] [INFO] Listening at: http://127.0.0.1:8000 (23570)
[2023-07-19 13:55:59 +0200] [23570] [INFO] Using worker: gthread
[2023-07-19 13:55:59 +0200] [23573] [INFO] Booting worker with pid: 23573
[2023-07-19 13:55:59 +0200] [23575] [INFO] Booting worker with pid: 23575
[2023-07-19 13:55:59 +0200] [23576] [INFO] Booting worker with pid: 23576

```

<img width="693" alt="Capture d’écran 2023-07-19 à 13 56 43" src="https://github.com/benoitc/gunicorn/assets/14498/158d567d-4415-4d4f-8be2-8848c7ceed00">

**Comment by wind-shift** [comment]
Hello @benoitc , thanks a lot for the fix.

I'll deploy it to production. Would you mind explaining the rationale behind this patch? I'm still curious about it, and above all, keep up the good work :)

**Reply by benoitc** [reply]
@dkr-sahar There is a work in progress to make the thread worker more efficient. The original idea was to considerate the initial connection idle, but turn out this change was making the worker more fragile in some conditions under load. I will revisit this feature in a future version.

**Comment by TurnrDev** [comment]
Surely as this is just a bug fix it should have been  a patch incrementation? What is the versioning scheme for gunicorn? I'm afraid to update any of my commercial projects in fear of breaking them with this shoddy release schedule

**Comment by benoitc** [comment]
it's been incremented like thid to signal others this version is not minor.

**Reply by TurnrDev** [reply]
Hi, I'd love to know more about your versioning system then. Is it not major.minor.patch?

**Reply by benoitc** [reply]
I will rephrase, this version has been incremented to enlighten the need to upgrade for it. We followin this schema since many years before someone decided unilateraly to relase a website about a versioning schema.

**Reply by TurnrDev** [reply]
My worry stems from 4 releases to 2 days, one of which lists "miscellaneous fixes" - how can I be sure this will still work now I expect it to work?

Also, this has had a major and a minor version release, which to my understanding indicates a major change to the API interface (I would presume to the command Argos or contig Argos since wsgi is a set standard) and a new feature, but all of the logs in the changelogs mention just fixes.

There is one log mentioning a jump in a major version to indicate the start of a new release cycle and there would be more information explaining, but I can't find this anywhere.

Essentially, I'm just trying to determine if these changes are safe for me to push to production and right now, I'm not sure.

**Reply by benoitc** [reply]
And the will be another release coming month. Master is supposed to be always stable until it isn't. Releases are picked from it. Unfortunately a lot people are now passively waiting a release to test. Which increase the chance to keep some issues hidden. 

Anyway to answer to your question, the patch version is really only used for internal versionning or version needed for deployments. Only X.Y should be relevant. Y may contain fixes or backward compatible changes. Time between them can be short, like it always had. I will make it clear in the doc. For what it worth, I am my self using latest Gunicorn in production.

**Reply by benoitc** [reply]
To add to above. I will also clarify the way we expect to make the next releases. To keep a more steady pace.

**Reply by TurnrDev** [reply]
Thank you for clarifying that. And 21.x didn't introduce any breaking changes, no?

**Reply by DavidOliver** [reply]
I've just updated to 21.2.0 from 20.1.0 for an old Wagtail 2/Django 3 site, and nothing seems broken.
